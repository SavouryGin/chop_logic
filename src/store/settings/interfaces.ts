import { Language } from 'types';

export interface SettingsInitialState {
  language: Language;
  flags: SettingsFlags;
}

export interface SettingsFlags {
  isNavigationOpen: boolean;
  isSidebarOpen: boolean;
  isDarkMode: boolean;
  isFullScreen: boolean;
  isSettingOpen: boolean;
  isSoundsEnabled: boolean;
  isMenuAnimationActive: boolean;
  isSidebarAnimationActive: boolean;
  isModalWindowClosingAnimationActive: boolean;
}

export type SettingsFlag = keyof SettingsFlags;
