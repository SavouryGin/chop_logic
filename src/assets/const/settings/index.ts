import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
};
