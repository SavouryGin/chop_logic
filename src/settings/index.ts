import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const languageStringOptions = ['en', 'ru'];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  linkedIn: 'https://www.linkedin.com/in/dmitrii-suroviagin/',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
  facebook: 'https://www.facebook.com/dmitrii.suroviagin',
};
