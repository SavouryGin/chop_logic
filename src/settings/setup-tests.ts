import crypto from 'crypto';
import '@testing-library/jest-dom/extend-expect';

Object.defineProperty(globalThis, 'crypto', {
  value: {
    randomUUID: () => crypto.randomUUID(),
  },
});
