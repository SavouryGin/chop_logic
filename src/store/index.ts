import createSagaMiddleware from 'redux-saga';
import rootReducer from './root-reducer';
import rootSaga from './root-saga';
import { configureStore } from '@reduxjs/toolkit';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([sagaMiddleware]),
});

sagaMiddleware.run(rootSaga);
