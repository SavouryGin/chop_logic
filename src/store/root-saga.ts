import propositionsDPSagas from './propositions/direct-proofs/sagas';
import propositionsNPSagas from './propositions/natural-proofs/sagas';
import truthTableSagas from './propositions/truth-tables/sagas';
import { all } from 'redux-saga/effects';

function* rootSaga(): Generator {
  yield all([propositionsDPSagas(), propositionsNPSagas(), truthTableSagas()]);
}

export default rootSaga;
