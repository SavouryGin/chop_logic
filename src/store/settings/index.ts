import { Language } from 'types';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { SettingsFlag, SettingsInitialState } from './interfaces';

export const settingsInitialState: SettingsInitialState = {
  language: 'en',
  flags: {
    isNavigationOpen: false,
    isSidebarOpen: false,
    isDarkMode: false,
    isFullScreen: false,
    isSettingOpen: false,
    isSoundsEnabled: false,
  },
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState: settingsInitialState,
  reducers: {
    resetSettings: () => {
      return settingsInitialState;
    },

    toggleFlag: (state, action: PayloadAction<SettingsFlag>) => {
      const currentFlagValue = state.flags[action.payload];
      state.flags[action.payload] = !currentFlagValue;
    },

    setLanguage: (state, action: PayloadAction<Language>) => {
      state.language = action.payload;
    },

    setFullScreenFlag: (state, action: PayloadAction<boolean>) => {
      state.flags.isFullScreen = action.payload;
    },

    setUpFlag: (state, action: PayloadAction<{ flag: SettingsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },
  },
});

export const settingsActions = { ...settingsSlice.actions };
