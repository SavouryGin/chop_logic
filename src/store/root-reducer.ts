import { combineReducers } from '@reduxjs/toolkit';
import { propositionsDPSlice } from './propositions/direct-proofs';
import { propositionsNPSlice } from './propositions/natural-proofs';
import { settingsSlice } from './settings/slice';

const rootReducer = combineReducers({
  settings: settingsSlice.reducer,
  propositionsDP: propositionsDPSlice.reducer,
  propositionsNP: propositionsNPSlice.reducer,
});

export default rootReducer;
