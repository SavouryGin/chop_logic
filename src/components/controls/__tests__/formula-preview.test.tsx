import FormulaPreview from '../formula-preview';
import React from 'react';
import propositionsFormulasItems from '__mocks__/data/propositions/formulas-items';
import propositionsTableItems from '__mocks__/data/propositions/table-items';
import renderWithRedux from 'helpers/test-utils/render-with-redux';
import { combineReducers } from '@reduxjs/toolkit';
import { settingsInitialState, settingsSlice } from 'store/settings';

describe('FormulaPreview component:', () => {
  const mockedReducer = combineReducers({ settings: settingsSlice.reducer });
  const mockedState = { settings: settingsInitialState };

  it('should render the formula preview', () => {
    const { asFragment } = renderWithRedux(
      <FormulaPreview preview={propositionsFormulasItems.propositionalExpression} />,
      mockedReducer,
      mockedState,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the error message', () => {
    const { asFragment } = renderWithRedux(<FormulaPreview preview={propositionsTableItems.error} />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
