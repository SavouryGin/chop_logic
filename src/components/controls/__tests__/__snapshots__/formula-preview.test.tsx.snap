// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormulaPreview component: should render the error message 1`] = `
<DocumentFragment>
  <div
    class="formula-preview"
  >
    <label
      class="label"
      data-testid="label-test-id"
    >
      <span
        class="label__text"
      >
        Preview
      </span>
    </label>
    <p
      class="formula-preview__error"
    >
      The entered expression cannot be converted into a propositional formula.
    </p>
  </div>
</DocumentFragment>
`;

exports[`FormulaPreview component: should render the formula preview 1`] = `
<DocumentFragment>
  <div
    class="formula-preview"
  >
    <label
      class="label"
      data-testid="label-test-id"
    >
      <span
        class="label__text"
      >
        Preview
      </span>
    </label>
    <pre
      class="formula formula-preview__formula"
      role="math"
    >
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <span
        class="formula__operator"
      >
        ¬
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <var
        class="formula__variable"
      >
        P
      </var>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__operator"
      >
         ⇒ 
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <var
        class="formula__variable"
      >
        P
      </var>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__operator"
      >
         ∧ 
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <var
        class="formula__variable"
      >
        Q
      </var>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__operator"
      >
         ⇒ 
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <span
        class="formula__operator"
      >
        ¬
      </span>
      <span
        class="formula__parentheses"
      >
        (
      </span>
      <var
        class="formula__variable"
      >
        P
      </var>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__parentheses"
      >
        )
      </span>
      <span
        class="formula__parentheses"
      >
        )
      </span>
    </pre>
  </div>
</DocumentFragment>
`;
